db/src/Pos/DB/Block/GState/BlockExtra.hs

```diff
@@ -60,8 +69,69 @@ isBlockInMainChain h =
 -- less than 'blkSecurityParam'.
 getLastSlots :: forall m . MonadDBRead m => m LastBlkSlots
 getLastSlots =
-    maybeThrow (DBMalformed "Last slots not found in the global state DB") =<<
-    gsGetBi lastSlotsKey
+    gsGetBi lastSlotsKey2 >>=
+        maybeThrow (DBMalformed "Last slots v2 not found in the global state DB")
+
+putLastSlots :: forall m . MonadDB m => LastBlkSlots -> m ()
+putLastSlots =
+    gsPutBi lastSlotsKey2
+
+-- | Roll back slots.
+-- This assumes that the tip header (retrieved via `getTipHeader` has already
+-- been rolled back.
+rollbackLastSlots :: forall m . (CanLog m, MonadIO m, MonadDB m) => Genesis.Config -> m ()
+rollbackLastSlots genesisConfig = do
+    -- Simple is better. Find the 'FlatSlotId' of the tip header and then generate
+    -- the required 'LastBlkSlots' data from that.
+    gsPutBi lastSlotsKey2 =<< getLastSlotInfo genesisConfig
+
+
+-- | This function acts as a one time conversion from version 1 to version 2
+-- of the `LastBlkSlots` data type.
+upgradeLastSlotsVersion :: forall m . (MonadDB m, MonadIO m) => Genesis.Config -> m ()
+upgradeLastSlotsVersion genesisConfig =
+    gsGetBi oldLastSlotsKey >>= \case
+        Nothing -> pure () -- Assume it has already been converted.
+        Just (_ :: OldestFirst [] FlatSlotId) -> do
+            gsPutBi lastSlotsKey2 =<< getLastSlotInfo genesisConfig
+            gsDelete oldLastSlotsKey -- Delete the old key
+  where
+    -- This is the DB key for version 1 of the `LastBlksSlots` data type that only
+    -- contains a list of `FlatSlotId`s
+    oldLastSlotsKey :: ByteString
+    oldLastSlotsKey = "e/ls/"
+
+
+-- Get the 'LastSlotInfo' data for the last 'k' (security paramenter) starting
+-- at the current blockchain tip.
+getLastSlotInfo :: (MonadDB m, MonadIO m) => Genesis.Config -> m LastBlkSlots
+getLastSlotInfo genesisConfig = do
+    th <- getTipHeader
+    let thfsid = flattenEpochOrSlot (configEpochSlots genesisConfig) $ getEpochOrSlot th
+        lastfsid =
+            if thfsid <= fromIntegral (configBlkSecurityParam genesisConfig - 1)
+                then 0
+                else thfsid - fromIntegral (configBlkSecurityParam genesisConfig - 1)
+    OldestFirst <$> convert th lastfsid []
+  where
+    convert :: (MonadDB m, MonadIO m) => BlockHeader -> FlatSlotId -> [LastSlotInfo] -> m [LastSlotInfo]
+    convert bh lastFsid !acc = do
+        let bhFsid = flattenEpochOrSlot (configEpochSlots genesisConfig) $ getEpochOrSlot bh
+        if bhFsid < lastFsid
+            then pure acc
+            else do
+                let ys = case leaderKey bh of
+                            Nothing -> acc
+                            Just lk -> LastSlotInfo bhFsid lk : acc
+                mnbh <- getHeader $ view prevBlockL bh
+                case mnbh of
+                    Nothing  -> pure acc
+                    Just nbh -> convert nbh lastFsid ys
```

Comment:
👍 If I understand correctly, `mnbh` would only be `Nothing` for the true genesis block (e0s0)